walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,145 @@
+    long requestDuration = request.getDuration();
+    int numRequestSlots = (int) requestDuration / 30;
+    int numEvents = events.size();
+
+    // if the duration is longer than a day there are no options
+    if (requestDuration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // if there are no attendees or events, the time range is the whole day
+    if (events.isEmpty() || allAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    // sets up timeSlots array; anywhere left true has all the requestAttendees available
+    boolean[] mandatoryTimeSlots = convertToTimeSlots(requestAttendees, events);
+    boolean[] allTimeSlots = convertToTimeSlots(allAttendees, events);    
+
+
+    ArrayList<Integer> possibleSlots = convertToSlots(mandatoryTimeSlots, numRequestSlots);
+    ArrayList<Integer> allSlots = convertToSlots(allTimeSlots, numRequestSlots);
+    
+    ArrayList<TimeRange> possibleTimes = convertToTimes(possibleSlots, numRequestSlots, numEvents);
+    ArrayList<TimeRange> allTimes = convertToTimes(allSlots, numRequestSlots, numEvents);
+
+    if (allTimes.isEmpty() && !possibleTimes.contains(TimeRange.WHOLE_DAY)) {
+        return possibleTimes;
+    }
+    return allTimes;
+  }
+  
+  public static boolean[] convertToTimeSlots (Collection<String> requestAttendees, Collection<Event> events) {
+    boolean[] timeSlots = new boolean[TOTAL_NUM_SLOTS];
+    Arrays.fill(timeSlots, Boolean.TRUE);
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      TimeRange when = event.getWhen();
+      int start = (when.start()) / 30;
+      int eventSlots = when.duration() / 30;
+
+      Iterator<String> eventAi = eventAttendees.iterator(); 
+      while(eventAi.hasNext()) {
+        String eventAttendee = eventAi.next();
+
+        Iterator<String> requestAi = requestAttendees.iterator(); 
+        while (requestAi.hasNext()) {
+          String requestAttendee = requestAi.next();
+          if (eventAttendee.equals(requestAttendee)) {
+            for (int i = 0; i < eventSlots; i++) {
+              timeSlots[start + i] = false;
+            }
+            break;
+          }
+        }
+      }
+    }
+    return timeSlots;
+  }
+
+  // converts boolean availability to list of available slots
+  public static ArrayList<Integer> convertToSlots (boolean timeSlots[], int numRequestSlots) {
+    ArrayList<Integer> slots = new ArrayList<Integer>();
+    boolean add;
+    for (int i = 0; i < TOTAL_NUM_SLOTS; i++) {
+      add = false;
+      if (timeSlots[i]) {
+        add = true;
+        // checks if the duration is longer than a single time slot
+        for (int j = 1; j < numRequestSlots; j++) {
+          if ((i+j) < timeSlots.length && !timeSlots[i+j]) {
+            add = false;
+          }
+        }
+      }
+      if (add) {
+        for (int j = 0; j < numRequestSlots; j++) {
+          slots.add(i+j);
```

Comment:
So, I think if I'm understanding this correctly, `slots` will contain essentially all the indexes (from `timeSlots`) that are potential answers. Would it help you to be creating a list of TimeRanges at this step? Could you potentially combine `convertToSlots` with `convertToTimes`? (just brainstorming here at how you could do this even more efficiently, maybe I'm completely missing something :) )

Reply:
I could probably combine them if I thought about it hard enough but it makes most sense to me when they're separate. So I'm just going to leave them as is.